src/mc-fle2-rfds.c

```diff
@@ -0,0 +1,615 @@
+      char *_str = bson_as_canonical_extended_json (bson, NULL); \
+      CLIENT_ERR (fmt ": %s", __VA_ARGS__, _str);                \
+      bson_free (_str);                                          \
+   } else                                                        \
+      (void) 0
+
+// Parses a document like {$and: []} and outputs an iterator to the array.
+static bool
+parse_and (const bson_t *in, bson_iter_t *out, mongocrypt_status_t *status)
+{
+   BSON_ASSERT_PARAM (in);
+   BSON_ASSERT_PARAM (out);
+   BSON_ASSERT (status || true);
+
+   bson_iter_t and;
+   if (!bson_iter_init (&and, in) || !bson_iter_next (&and) ||
+       0 != strcmp (bson_iter_key (&and), "$and")) {
+      ERR_WITH_BSON (in, "%s", "error unable to find '$and'");
+      return false;
+   }
+
+   if (!BSON_ITER_HOLDS_ARRAY (&and)) {
+      ERR_WITH_BSON (in, "%s", "expected '$and' to be array");
+      return false;
+   }
+
+   *out = and;
+
+   if (bson_iter_next (&and)) {
+      ERR_WITH_BSON (
+         in, "unexpected extra key '%s' after '$and'", bson_iter_key (&and));
+      return false;
+   }
+   return true;
+}
+
+typedef struct {
+   const char *op_type_str;
+   const char *field;
+   bson_iter_t value;
+   mc_FLE2RangeOperator_t op_type;
+} operator_value_t;
+
+// Parses a document like {$gt: ["$age", 5]}.
+static bool
+parse_aggregate_expression (const bson_t *orig,
+                            bson_iter_t in,
+                            operator_value_t *out,
+                            mongocrypt_status_t *status)
+{
+   BSON_ASSERT_PARAM (orig);
+   BSON_ASSERT_PARAM (out);
+   BSON_ASSERT (status || true);
+
+   bson_iter_t array, value;
+   const char *op_type_str = bson_iter_key (&in);
+   bool ok = false;
+   const char *field;
+
+   if (!BSON_ITER_HOLDS_ARRAY (&in)) {
+      ERR_WITH_BSON (orig, "%s", "expected argument to be array");
+      goto fail;
+   }
+
+   if (!bson_iter_recurse (&in, &array)) {
+      ERR_WITH_BSON (orig, "%s", "failed to recurse into array");
+      goto fail;
+   }
+
+   // Expect exactly 2 elements, like ["$age", 5]. The first element is the
+   // field path. The second element is the value.
+   if (!bson_iter_next (&array)) {
+      ERR_WITH_BSON (
+         orig, "expected 2 elements in operand %s, got 0", op_type_str);
+      goto fail;
+   }
+   if (!BSON_ITER_HOLDS_UTF8 (&array)) {
+      ERR_WITH_BSON (
+         orig, "expected UTF-8 as first element in %s", op_type_str);
+      goto fail;
+   }
+   field = bson_iter_utf8 (&array, NULL);
+
+   if (!bson_iter_next (&array)) {
+      ERR_WITH_BSON (
+         orig, "expected 2 elements in operand %s, got 1", op_type_str);
+      goto fail;
+   }
+   value = array;
+   if (bson_iter_next (&array)) {
+      ERR_WITH_BSON (
+         orig, "expected 2 elements in operand %s, got > 2", op_type_str);
+      goto fail;
+   }
+
+   out->field = field;
+   out->op_type_str = op_type_str;
+   out->op_type = get_operator_type (op_type_str);
+   out->value = value;
```

Comment:
```suggestion
   *out = (operator_value_t){
      .field = field,
      .op_type_str = op_type_str,
      .op_type = get_operator_type (op_type_str),
      .value = value,
   };

```
Opt into `-Wmissing-field-initializers` warnings.